### DISCUSSION TITLE
procedure for generating a zfs-2.2.0-rcX feature flags compatible zfsbootmenu EFI image?

### DISCUSSION THREAD

**Discussion by comrade-meowski** [zbm-dev/zfsbootmenu#456] [Questions]
Hi, I can't seem to coax my Ubuntu/Arch test VMs into doing this: I'm pretty familiar with zfsbootmenu and have read through the documentation but I'm still missing something.

Specifically I'd like to generate an updated zbm.efi from a pre-existing kernel with zfs-2.2.0-rcX that can boot a rpool with the new feature flags enabled. I've experimented with generate-zbm and zbm-builder.sh without success. 

Anyone already done this?

**Comment by zdykstra** [comment]
You'll need to create a Void package for an updated ZFS and then add it to the build container. Otherwise if you build directly on the running system, Dracut/mkinitcpio will use the kernel modules that are installed for the specified kernel.

Is ZFSBootMenu with ZFS 2.1.12 unable to read a pool with  2.2.0 feature flags? It will detect and alert which flags are incompatible. Though historically older versions of ZFS can still import newer versions read-only (which is the default behavior for ZBM).

**Comment by comrade-meowski** [comment]
Thanks for your response - I have now solved this issue. Following the path of least resistance:

Noped the zbm-builder.sh method immediately - figuring out how to modify the container logic and  learn how to package zfs-2.2.0-rc3 in xpbs for Void looked... unappetizing. 

Ubuntu was a non-starter for generate-zbm: it depends on mkinitcpio/dracut and both are essentially foreign packages on that distro.  I should have remembered that earlier.

Arch on the other hand does have native mkinitcpio/dracut so was the logical choice - it just needed a zfs-2.2.0-rc3 package installed for generate-zbm to build an EFI from. That doesn't exist so I made a PKGBUILD as that was a lot easier than any of the other options. 

So Arch + kernel 6.4.7 + zfs-2.2.0-rc3 + generate-zbm (git master) = new.zbm.efi

new.zbm.efi is of course portable and correctly boots zfs-2.2.0 rpools with all the new feature flags enabled on the Ubuntu/Arch/Fedora VMs I've tested it on. 

Many thanks for your excellent software @zdykstra - I shall mark this closed.

**Reply by zdykstra** [reply]
`dracut-core` can be used on Debian/Ubuntu without conflicts or replacing the native initramfs generator.

Which specific ZFS 2.2.0 feature flags were causing you problems?

**Comment by comrade-meowski** [comment]
The new feature flags that zfs 2.1.12 > 2.2.0-rcX brings are:

      zilsaxattr
      head_errlog
      blake3
      block_cloning
      vdev_zaps_v2

Enabling all of them on a rpool requires building a new zbm.efi (unsurprisingly). 

For completeness sake I'll take another look at dracut-core on Ubuntu but I'm perfectly happy with the current solution.

**Reply by zdykstra** [reply]
For the record, `com.delphix:head_errlog` is the actual feature flag that isn't backwards compatible with ZFS 2.1.12 in both read-only and read-write mode. `org.openzfs:zilsaxattr` blocks importing a pool read-write. We'll be cutting a new release of ZFSBootMenu as soon as ZFS 2.2.0 is tagged and released.
